src/main/java/io/leonis/subra/ipc/serialization/protobuf/vision/GoalsDeducer.java

```diff
@@ -0,0 +1,88 @@
+package io.leonis.subra.ipc.serialization.protobuf.vision;
+
+import java.util.Set;
+import java.util.stream.*;
+import org.nd4j.linalg.factory.Nd4j;
+import org.reactivestreams.Publisher;
+import io.leonis.subra.game.data.*;
+import io.leonis.zosma.game.engine.Deducer;
+import io.leonis.algieba.geometry.CardinalDirection;
+import org.robocup.ssl.Geometry.*;
+import org.robocup.ssl.Wrapper.WrapperPacket;
+import reactor.core.publisher.Flux;
+
+/**
+ * The Class GoalsDeducer
+ *
+ * Reads all goaldata from a {@link GeometryFieldSize}.
+ *
+ * @author Jeroen de Jong
+ */
+public class GoalsDeducer implements Deducer<WrapperPacket, Set<Goal>> {
+  @Override
+  public Publisher<Set<Goal>> apply(final Publisher<WrapperPacket> geometryPublisher) {
+    return Flux.from(geometryPublisher)
+        .map(WrapperPacket::getGeometry)
+        .map(GeometryData::getField)
+        .map(input -> Stream.of(input)
+            .flatMap(packet -> Stream.of(
+                this.createGoal(TeamColor.BLUE, CardinalDirection.NORTH, packet),
+                this.createGoal(TeamColor.YELLOW, CardinalDirection.SOUTH, packet)))
+            .collect(Collectors.toSet()));
+  }
+
+  private Goal createGoal(
+      final TeamColor teamColor,
+      final CardinalDirection direction,
+      final GeometryFieldSize geometryData
+  ) {
+    switch (direction) {
+      case NORTH:
```

Comment:
I still dislike this tbh 😓  Just the x andy changes right? can't we determine those two floats in the switch case and construct the object in the end?